File: server/models/transaction.go
package models

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type Transaction struct {
	ID         uuid.UUID `gorm:"column:id;type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	Type       Status    `gorm:"column:type;not null;type:varchar(20)" json:"type"` // Deposit / Withdraw
	Amount     uint      `gorm:"column:amount;not null" json:"amount"`
	AdminID    uuid.UUID `gorm:"column:admin_id;type:uuid;index" json:"admin_id"`
	InvestorID uuid.UUID `gorm:"column:investor_id;type:uuid;index" json:"investor_id"`
	CreatedAt  time.Time `gorm:"column:created_at;autoCreateTime" json:"created_at"`

	Admin    *Admin    `gorm:"foreignKey:AdminID" json:"admin"`
	Investor *Investor `gorm:"foreignKey:InvestorID" json:"investor"`
}

func (Transaction) TableName() string {
	return "transaction"
}

func (t Transaction) String() string {
	str := fmt.Sprintf("Transaction(id=%s, type='%s', amount=%d, admin_id='%s', investor_id='%s'", t.ID, t.Type, t.Amount, t.AdminID, t.InvestorID)
	str += ")"
	return str
}

File: client/src/models/Transaction.ts
import type Admin from "./Admin";

type TransactionJSON = {
  id?: string;
  type?: string;
  amount?: number;
  admin_id?: string;
  investor_id?: string;
  created_at?: string;
};
class Transaction {
  #id: string;
  type: "deposit" | "withdraw" | "unnkown";
  amount: number;
  #admin_id: string;
  #investor_id: string;
  #created_at: string;
  #admin_ref: Admin | null;
  constructor(trJSON: TransactionJSON, admin?: Admin) {
    this.#id = trJSON.id || "";
    this.amount = trJSON.amount || 0;
    this.#admin_id = trJSON.admin_id || "";
    this.#investor_id = trJSON.investor_id || "";
    this.#created_at = trJSON.created_at || "";
    this.#admin_ref = admin || null;
    switch (trJSON.type) {
      case "deposit":
        this.type = "deposit";
        break;
      case "withdraw":
        this.type = "withdraw";
        break;

      default:
        this.type = "unnkown";
        break;
    }
  }
  get id(): string {
    return this.#id;
  }
  get admin(): Admin | null {
    return this.#admin_ref;
  }
  get admin_id(): string {
    return this.#admin_id;
  }
  get investor_id(): string {
    return this.#investor_id;
  }
  get created_at(): string {
    return this.#created_at;
  }
  get date(): Date {
    return new Date(this.#created_at);
  }
}

export { type TransactionJSON };
export default Transaction;
